File: Battleship.Logic/GameEngine.cs
using Battleship.Shared;
namespace Battleship.Logic;
public class GameEngine : IGameEngine
{
    private readonly Board[] _playerBoards;
    private readonly ShipType[] _standardShips = 
    {
        ShipType.Carrier,
        ShipType.Battleship,
        ShipType.Destroyer,
        ShipType.Submarine,
        ShipType.PatrolBoat
    };
    public GameEngine()
    {
        _playerBoards = new Board[2] { new Board(), new Board() };
    }
    public Board GetPlayerBoard(int player) => _playerBoards[player - 1];
    public Board GetOpponentBoard(int player) => _playerBoards[player == 1 ? 1 : 0];
    public bool PlaceShip(ShipType type, Coordinate start, Orientation orientation)
    {
        var board = GetPlayerBoard(CurrentPlayer);
        return board.PlaceShip(type, start, orientation);
    }
    public bool PlaceShipForPlayer(ShipType type, Coordinate start, Orientation orientation, int player)
    {
        var board = GetPlayerBoard(player);
        return board.PlaceShip(type, start, orientation);
    }
    public ShotResult FireShot(Coordinate target)
    {
        var opponentBoard = GetOpponentBoard(CurrentPlayer);
        var result = opponentBoard.FireShot(target);
        if (opponentBoard.AreAllShipsSunk())
        {
        }
        else
        {
            SwitchPlayer();
        }
        return result;
    }
    public bool IsGameOver => _playerBoards.Any(board => board.AreAllShipsSunk());
    public int CurrentPlayer { get; private set; } = 1;
    public void SwitchPlayer()
    {
        CurrentPlayer = CurrentPlayer == 1 ? 2 : 1;
    }
    public bool HasPlayerCompletedPlacement(int player)
    {
        var board = GetPlayerBoard(player);
        var requiredShips = _standardShips.Length;
        return board.Ships.Count() == requiredShips;
    }
    public List<ShipType> GetRemainingShips(int player)
    {
        var board = GetPlayerBoard(player);
        var placedShipTypes = board.Ships.Select(s => s.Type).ToHashSet();
        return _standardShips.Where(shipType => !placedShipTypes.Contains(shipType)).ToList();
    }
    public bool CanPlaceShip(ShipType type, Coordinate start, Orientation orientation, int player)
    {
        var board = GetPlayerBoard(player);
        return board.CanPlaceShip(type, start, orientation);
    }
    public List<Ship> GenerateRandomPlacement(int player)
    {
        var ships = new List<Ship>();
        var random = new Random();
        var shipTypes = (ShipType[])_standardShips.Clone();
        foreach (var shipType in shipTypes)
        {
            Ship ship;
            int attempts = 0;
            do
            {
                var start = new Coordinate(random.Next(10), random.Next(10));
                var orientation = random.Next(2) == 0 ? Orientation.Horizontal : Orientation.Vertical;
                ship = new Ship(shipType, start, orientation);
                attempts++;
            } while (!CanPlaceShip(shipType, ship.Start, ship.Orientation, player) && attempts < 100);
            if (attempts < 100)
            {
                PlaceShipForPlayer(shipType, ship.Start, ship.Orientation, player);
                ships.Add(ship);
            }
            else
            {
                return new List<Ship>();
            }
        }
        return ships;
    }
    public void ResetPlayerBoard(int player)
    {
        _playerBoards[player - 1] = new Board();
    }
}

File: battleships-go/internal/logic/gameengine.go
package logic

import (
	"battleships-go/internal/network"
	"battleships-go/pkg/models"
	"math/rand"
	"time"
)

type GameEngine struct {
	playerBoards  [2]*network.Board
	standardShips []models.ShipType
	currentPlayer int
}

func NewGameEngine() *GameEngine {
	return &GameEngine{
		playerBoards: [2]*network.Board{
			network.NewBoard(),
			network.NewBoard(),
		},
		standardShips: []models.ShipType{
			models.Carrier,
			models.Battleship,
			models.Destroyer,
			models.Submarine,
			models.PatrolBoat,
		},
		currentPlayer: 1,
	}
}

func (e *GameEngine) GetPlayerBoard(player int) *network.Board {
	return e.playerBoards[player-1]
}

func (e *GameEngine) GetOpponentBoard(player int) *network.Board {
	if player == 1 {
		return e.playerBoards[1]
	}
	return e.playerBoards[0]
}

func (e *GameEngine) CurrentPlayer() int {
	return e.currentPlayer
}

func (e *GameEngine) PlaceShipForPlayer(shipType models.ShipType, start models.Coordinate, orientation models.Orientation, player int) bool {
	board := e.GetPlayerBoard(player)
	return board.PlaceShip(shipType, start, orientation)
}

func (e *GameEngine) FireShot(target models.Coordinate) models.ShotResult {
	opponentBoard := e.GetOpponentBoard(e.currentPlayer)
	result := opponentBoard.FireShot(target)

	if !opponentBoard.AreAllShipsSunk() {
		e.SwitchPlayer()
	}
	return result
}

func (e *GameEngine) IsGameOver() bool {
	for _, board := range e.playerBoards {
		if board.AreAllShipsSunk() {
			return true
		}
	}
	return false
}

func (e *GameEngine) SwitchPlayer() {
	if e.currentPlayer == 1 {
		e.currentPlayer = 2
	} else {
		e.currentPlayer = 1
	}
}

func (e *GameEngine) HasPlayerCompletedPlacement(player int) bool {
	board := e.GetPlayerBoard(player)
	return len(board.Ships) == len(e.standardShips)
}

func (e *GameEngine) GenerateRandomPlacement(player int) []*models.Ship {
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	var placedShips []*models.Ship

	for _, shipType := range e.standardShips {
		success := false
		for attempts := 0; attempts < 100; attempts++ {
			start := models.Coordinate{X: r.Intn(10), Y: r.Intn(10)}
			orientation := models.Orientation(r.Intn(2))

			if e.playerBoards[player-1].CanPlaceShip(shipType, start, orientation) {
				e.PlaceShipForPlayer(shipType, start, orientation, player)
				placedShips = append(placedShips, &models.Ship{
					Type:        shipType,
					Start:       start,
					Orientation: orientation,
				})
				success = true
				break

			}
		}
		if !success {
			return nil
		}
	}
	return placedShips
}

func (e *GameEngine) ResetPlayerBoard(player int) {
	e.playerBoards[player-1] = network.NewBoard()
}
